File: algorithms/239.Sliding_Window_Maximum/Sliding_Window_Maximum.py
from collections import deque

class Solution:
    # @param {integer[]} nums
    # @param {integer} k
    # @return {integer[]}
    def maxSlidingWindow(self, nums, k):
        q = deque()
        max_numbers = []

        for i in xrange(k):
            while q and nums[i] >= nums[q[-1]]:
                q.pop()
            q.append(i)

        for i in xrange(k, len(nums)):
            max_numbers.append(nums[q[0]])

            while q and nums[i] >= nums[q[-1]]:
                q.pop()

            while q and q[0] <= i - k:
                q.popleft()

            q.append(i)

        if q:
            max_numbers.append(nums[q[0]])

        return max_numbers


from collections import deque
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        if not nums: return []
        res = []
        dq = deque()  # store index
        for i in xrange(len(nums)):
            if dq and dq[0]<i-k+1:  # out of the window
                dq.popleft()
            while dq and nums[dq[-1]]<nums[i]:  # remove impossible candidate
                dq.pop()
            dq.append(i)
            if i>k-2:
                res.append(nums[dq[0]])
        return res

File: algorithms/239.Sliding_Window_Maximum/Sliding_Window_Maximum.java
<<<<<<< HEAD
public class Solution {
    public int[] maxSlidingWindow(int[] a, int k) {
     if (a == null || k <= 0) return new int[0];         
     int[] res = new int[a.length - k + 1];
     ArrayDeque<Integer> deque = new ArrayDeque<Integer>(); 

     int index  = 0;
     for (int i = 0; i < a.length; i++) { 
         while (!deque.isEmpty() && deque.peek() < i - k + 1) // Ensure deque's size doesn't exceed k
             deque.poll();

        // Remove numbers smaller than a[i] from right(a[i-1]) to left, to make the first number in the deque the largest one in the window      
         while (!deque.isEmpty() && a[deque.peekLast()] < a[i]) 
             deque.pollLast();

         deque.offer(i);// Offer the current index to the deque's tail

         if (i >= k - 1)// Starts recording when i is big enough to make the window has k elements 
             res[index++] = a[deque.peek()];
     }       
     return res;
 }
=======
public class Solution {
    public int[] maxSlidingWindow(int[] a, int k) {
     if (a == null || k <= 0) return new int[0];         
     int[] res = new int[a.length - k + 1];
     ArrayDeque<Integer> deque = new ArrayDeque<Integer>(); 

     int index  = 0;
     for (int i = 0; i < a.length; i++) { 
         while (!deque.isEmpty() && deque.peek() < i - k + 1) // Ensure deque's size doesn't exceed k
             deque.poll();

        // Remove numbers smaller than a[i] from right(a[i-1]) to left, to make the first number in the deque the largest one in the window      
         while (!deque.isEmpty() && a[deque.peekLast()] < a[i]) 
             deque.pollLast();

         deque.offer(i);// Offer the current index to the deque's tail

         if (i >= k - 1)// Starts recording when i is big enough to make the window has k elements 
             res[index++] = a[deque.peek()];
     }       
     return res;
 }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
}